File: ga/4omini_result_2024_7/result/multi_thread_basic_mini/6_multi_thread_basic.c
use std::sync::{Arc, Mutex};
use std::thread;

fn withdraw(account_balance: Arc<Mutex<i32>>, amount: i32) {
    for _ in 0..amount {
        let mut balance = account_balance.lock().unwrap();
        *balance -= 1;
    }
}

fn main() {
    let account_balance = Arc::new(Mutex::new(1_000_000)); // Initial account balance

    let account_clone1 = Arc::clone(&account_balance);
    let account_clone2 = Arc::clone(&account_balance);

    let t1 = thread::spawn(move || {
        withdraw(account_clone1, 500_000);
    });

    let t2 = thread::spawn(move || {
        withdraw(account_clone2, 500_000);
    });

    t1.join().unwrap();
    t2.join().unwrap();

    // Print the final account balance
    let final_balance = account_balance.lock().unwrap();
    println!("Final account balance: {}", final_balance);
}

File: ga/glm4_result_2024_6/result/multi_thread_basic_cpp_glm/6_multi_thread_basic.rs
use std::sync::{Arc, atomic::{AtomicI32, Ordering}};
use std::thread;

fn withdraw(account_balance: &Arc<AtomicI32>, amount: i32) {
    for _ in 0..amount {
        account_balance.fetch_sub(1, Ordering::SeqCst);
    }
}

fn main() {
    let account_balance = Arc::new(AtomicI32::new(1000000));

    let account_balance1 = account_balance.clone();
    let t1 = thread::spawn(move || withdraw(&account_balance1, 500000));

    let account_balance2 = account_balance.clone();
    let t2 = thread::spawn(move || withdraw(&account_balance2, 500000));

    t1.join().unwrap();
    t2.join().unwrap();

    println!("Final account balance: {}", account_balance.load(Ordering::SeqCst));
}